[CI] TextFieldMapperTests testSyntheticSourceMany failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #3185 / ubuntu-1804_platform-support-unix](https://gradle-enterprise.elastic.co/s/35repdvst5gbu)
- [elasticsearch-periodic-platform-support #3185 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/jdeqkumbbiqxu)
- [elasticsearch-periodic #3211 / single-processor-node-tests](https://gradle-enterprise.elastic.co/s/l7hkg5tefo6lc)
- [elasticsearch-periodic #3211 / concurrent-search-tests](https://gradle-enterprise.elastic.co/s/m74ukybvcpblq)
- [elasticsearch-periodic #3211 / graalvm-ce17_checkpart1_java-matrix](https://gradle-enterprise.elastic.co/s/kqjqx5oqmqxyc)
- [elasticsearch-pull-request #23744 / part-1](https://gradle-enterprise.elastic.co/s/5ym6aj2r33rtc)

**Reproduction Line:**
```
./gradlew ':server:test' --tests "org.elasticsearch.index.mapper.TextFieldMapperTests.testSyntheticSourceMany" -Dtests.seed=BA7BC64B128BD4E0 -Dtests.jvm.argline="-Des.concurrent_search=true" -Dtests.locale=mk -Dtests.timezone=America/Catamarca -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testSyntheticSourceMany'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.index.mapper.TextFieldMapperTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: doc 0
Expected: "{\"field\":[\"sXwGb\",\"VuZdEwGGuxZSstbReXqgKkoWpJfsKmkrNSgUNLtktZjgnXAvDyfxvnWJQxBqGKSHyUyeAbfZTcOSKXlOQrpNOOTPyqqNdNxSTZFtxJgmj\"]}"
     but: was "{\"field\":[\"VuZdEwGGuxZSstbReXqgKkoWpJfsKmkrNSgUNLtktZjgnXAvDyfxvnWJQxBqGKSHyUyeAbfZTcOSKXlOQrpNOOTPyqqNdNxSTZFtxJgmj\",\"sXwGb\"]}"
```

**Issue Reasons:**
- [main] 6 failures in test testSyntheticSourceMany (0.6% fail rate)

**Note:**
This issue was created using new, in-development muting automation. Please mind any issues.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -84,9 +84,6 @@ tests:
 - class: "org.elasticsearch.test.jvm_crash.JvmCrashIT"
   issue: "https://github.com/elastic/elasticsearch/issues/110052"
   method: "testJvmCrash"
-- class: org.elasticsearch.index.mapper.KeywordFieldMapperTests
-  method: testBlockLoaderFromRowStrideReader
-  issue: https://github.com/elastic/elasticsearch/issues/110163
 
 # Examples:
 #

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -84,9 +84,6 @@ tests:
 - class: "org.elasticsearch.test.jvm_crash.JvmCrashIT"
   issue: "https://github.com/elastic/elasticsearch/issues/110052"
   method: "testJvmCrash"
-- class: org.elasticsearch.index.mapper.TextFieldMapperTests
-  method: testSyntheticSourceMany
-  issue: https://github.com/elastic/elasticsearch/issues/110161
 - class: org.elasticsearch.index.mapper.KeywordFieldMapperTests
   method: testBlockLoaderFromRowStrideReader
   issue: https://github.com/elastic/elasticsearch/issues/110163

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/KeywordFieldSyntheticSourceSupport.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/KeywordFieldSyntheticSourceSupport.java
@@ -19,6 +19,7 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import static org.hamcrest.Matchers.equalTo;
 
@@ -62,11 +63,6 @@ public class KeywordFieldSyntheticSourceSupport implements MapperTestCase.Synthe
         List<Tuple<String, String>> values = ESTestCase.randomList(1, maxValues, this::generateValue);
         List<String> in = values.stream().map(Tuple::v1).toList();
 
-        if (preservesExactSource()) {
-            var blockLoaderResult = in.stream().filter(Objects::nonNull).toList();
-            return new MapperTestCase.SyntheticSourceExample(in, in, blockLoaderResult, this::mapping);
-        }
-
         List<String> validValues = new ArrayList<>();
         List<String> ignoredValues = new ArrayList<>();
         values.stream().map(Tuple::v2).forEach(v -> {
@@ -76,9 +72,17 @@ public class KeywordFieldSyntheticSourceSupport implements MapperTestCase.Synthe
                 validValues.add(v);
             }
         });
-
         List<String> outputFromDocValues = new HashSet<>(validValues).stream().sorted().collect(Collectors.toList());
-        List<String> syntheticSourceOutputList = store ? validValues : outputFromDocValues;
+
+        Object out;
+        if (preservesExactSource()) {
+            out = in;
+        } else {
+            var validValuesInCorrectOrder = store ? validValues : outputFromDocValues;
+            var syntheticSourceOutputList = Stream.concat(validValuesInCorrectOrder.stream(), ignoredValues.stream()).toList();
+            out = syntheticSourceOutputList.size() == 1 ? syntheticSourceOutputList.get(0) : syntheticSourceOutputList;
+        }
+
         List<String> loadBlock;
         if (loadBlockFromSource) {
             // The block loader infrastructure will never return nulls. Just zap them all.
@@ -86,12 +90,11 @@ public class KeywordFieldSyntheticSourceSupport implements MapperTestCase.Synthe
         } else if (docValues) {
             loadBlock = List.copyOf(outputFromDocValues);
         } else {
-            loadBlock = List.copyOf(syntheticSourceOutputList);
+            // Meaning loading from terms.
+            loadBlock = List.copyOf(validValues);
         }
-        Object loadBlockResult = loadBlock.size() == 1 ? loadBlock.get(0) : loadBlock;
 
-        syntheticSourceOutputList.addAll(ignoredValues);
-        Object out = syntheticSourceOutputList.size() == 1 ? syntheticSourceOutputList.get(0) : syntheticSourceOutputList;
+        Object loadBlockResult = loadBlock.size() == 1 ? loadBlock.get(0) : loadBlock;
         return new MapperTestCase.SyntheticSourceExample(in, out, loadBlockResult, this::mapping);
     }
 

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/TextFieldFamilySyntheticSourceTestSetup.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/TextFieldFamilySyntheticSourceTestSetup.java
@@ -117,10 +117,7 @@ public final class TextFieldFamilySyntheticSourceTestSetup {
             // Block loader will not use keyword multi-field if it has ignore_above configured.
             // And in this case it will use values from source.
             boolean loadingFromSource = ignoreAbove != null;
-            MapperTestCase.SyntheticSourceExample delegate = keywordMultiFieldSyntheticSourceSupport.example(
-                maxValues,
-                loadingFromSource
-            );
+            MapperTestCase.SyntheticSourceExample delegate = keywordMultiFieldSyntheticSourceSupport.example(maxValues, loadingFromSource);
 
             return new MapperTestCase.SyntheticSourceExample(
                 delegate.inputValue(),

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/TextFieldFamilySyntheticSourceTestSetup.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/TextFieldFamilySyntheticSourceTestSetup.java
@@ -84,11 +84,20 @@ public final class TextFieldFamilySyntheticSourceTestSetup {
         private final String fieldType;
         private final boolean store;
         private final boolean index;
+        private final Integer ignoreAbove;
+        private final KeywordFieldSyntheticSourceSupport keywordMultiFieldSyntheticSourceSupport;
 
         TextFieldFamilySyntheticSourceSupport(String fieldType, boolean supportsCustomIndexConfiguration) {
             this.fieldType = fieldType;
             this.store = randomBoolean();
             this.index = supportsCustomIndexConfiguration == false || randomBoolean();
+            this.ignoreAbove = randomBoolean() ? null : between(10, 100);
+            this.keywordMultiFieldSyntheticSourceSupport = new KeywordFieldSyntheticSourceSupport(
+                ignoreAbove,
+                randomBoolean(),
+                null,
+                false
+            );
         }
 
         @Override
@@ -105,18 +114,10 @@ public final class TextFieldFamilySyntheticSourceTestSetup {
                 return storedFieldExample(maxValues, mapping);
             }
 
-            Integer ignoreAbove = randomBoolean() ? null : between(10, 100);
-            var syntheticSourceSupportForKeywordMultiField = new KeywordFieldSyntheticSourceSupport(
-                ignoreAbove,
-                randomBoolean(),
-                null,
-                false
-            );
-
             // Block loader will not use keyword multi-field if it has ignore_above configured.
             // And in this case it will use values from source.
             boolean loadingFromSource = ignoreAbove != null;
-            MapperTestCase.SyntheticSourceExample delegate = syntheticSourceSupportForKeywordMultiField.example(
+            MapperTestCase.SyntheticSourceExample delegate = keywordMultiFieldSyntheticSourceSupport.example(
                 maxValues,
                 loadingFromSource
             );

